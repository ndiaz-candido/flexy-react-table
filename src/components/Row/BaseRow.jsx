import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  LeftSection,
  RightSection,
  ScrollableSection,
} from './sections';

const defaultRowStyle = {
  display: 'flex',
  overflow: 'hidden',
  minHeight: '50px',
};
const stickyStyle = {
  position: 'sticky',
  top: '0',
  zIndex: '1000',
};
const cellDefaultStyles = width => ({
  width: `${width}px`,
  flex: `0 0 ${width}px`,
});

class BaseRow extends Component {
  static DEFAULT_COLUMN_WIDTH = 50;
  static HEADER_ROW_ID = 0;
  static NOT_SPECIFY_COLUMN_VALUE = 'Not Column';

  state = {
    leftStickySection: [],
    rightStickySection: [],
    scrollableSection: [],
  };

  static getDerivedStateFromProps({ columns, rowData, columnClassName }) {
    const leftStickySection = [];
    const rightStickySection = [];
    const scrollableSection = [];

    columns.forEach(column => {
      const columnWidth = column.width || BaseRow.DEFAULT_COLUMN_WIDTH;
      const columnDataKey = column.dataKey || '';
      const cellContent = rowData[columnDataKey] || '';
      const styles = cellDefaultStyles(columnWidth);
      const cell = (
        <div className={columnClassName} style={styles}>
          {cellContent}
        </div>
      );

      switch (column.stickyAlign) {
        case 'left':
          leftStickySection.push(cell);
          break;
        case 'right':
          rightStickySection.push(cell);
          break;
        default:
          scrollableSection.push(cell);
          break;
      }
    });

    return {
      leftStickySection,
      rightStickySection,
      scrollableSection,
    };
  }

  render() {
    const { className, isSticky, onScroll, rowId, scrollLeft } = this.props;
    const {
      leftStickySection,
      rightStickySection,
      scrollableSection,
    } = this.state;
    const stickyStyleProps = isSticky ? stickyStyle : {};
    const styles = {
      ...defaultRowStyle,
      ...stickyStyleProps,
    };

    return (
      <div className={className} style={styles}>
        <LeftSection cells={leftStickySection} />
        <ScrollableSection
          rowId={rowId}
          cells={scrollableSection}
          onScroll={onScroll}
          scrollLeft={scrollLeft}
        />
        <RightSection cells={rightStickySection} />
      </div>
    );
  }
}

BaseRow.propTypes = {
  columns: PropTypes.array,
  className: PropTypes.any,
  columnClassName: PropTypes.any,
  isSticky: PropTypes.bool,
  isHeader: PropTypes.bool,
  onScroll: PropTypes.func,
  rowId: PropTypes.number,
  rowData: PropTypes.object,
  scrollLeft: PropTypes.number,
};

BaseRow.defaultProps = {
  columns: [],
  className: '',
  columnClassName: '',
  isSticky: false,
  isHeader: false,
  onScroll: () => {},
  rowId: -1,
  rowData: {},
  scrollLeft: 0,
};

export default BaseRow;
