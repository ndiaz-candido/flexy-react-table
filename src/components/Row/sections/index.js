import LeftSection from './LeftSection';
import RightSection from './RightSection';
import ScrollableSection from './ScrollableSection';

export { LeftSection, RightSection, ScrollableSection };
