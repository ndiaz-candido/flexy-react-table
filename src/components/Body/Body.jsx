import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import BaseRow from '../Row/BaseRow';

class Body extends Component {
  state = {
    scrollLeft: 0,
  };

  handleScrollEvent = ({ scrollLeft }) => {
    this.setState({ scrollLeft });
  };

  buildHeaderRowData = () => {
    const headerRowData = {};
    const { columns, columnHeaderRenderer } = this.props;

    columns.forEach((column) => {
      headerRowData[column.dataKey] =
        columnHeaderRenderer ? columnHeaderRenderer({ ...column }) : column.label;
    });

    return headerRowData;
  };

  render() {
    const {
      columns,
      columnClassName,
      headerRowClassName,
      headerColumnClassName,
      rows,
      rowClassName,
      stickHeader
    } = this.props;
    const { scrollLeft } = this.state;

    return (
      <Fragment>
        <BaseRow
          columns={columns}
          columnClassName={headerColumnClassName}
          className={headerRowClassName}
          isSticky={stickHeader}
          isHeader
          onScroll={this.handleScrollEvent}
          rowId={BaseRow.HEADER_ROW_ID}
          rowData={this.buildHeaderRowData()}
          scrollLeft={scrollLeft}
        />
        {rows.map((row, rowIndex) =>
          <BaseRow
            columns={columns}
            columnClassName={columnClassName}
            className={rowClassName}
            key={rowIndex}
            onScroll={this.handleScrollEvent}
            rowId={rowIndex + 1}
            rowData={row.props.rowData}
            scrollLeft={scrollLeft}
          />
        )}
      </Fragment>
    );
  }
}

Body.propTypes = {
  columns: PropTypes.arrayOf(PropTypes.object),
  columnHeaderRenderer: PropTypes.func,
  columnClassName: PropTypes.any,
  headerRowClassName: PropTypes.any,
  headerColumnClassName: PropTypes.any,
  rows: PropTypes.array.isRequired,
  rowClassName: PropTypes.any,
  stickHeader: PropTypes.bool,
};

Body.defaultProps = {
  columns: [],
  columnHeaderRenderer: null,
  columnClassName: '',
  headerRowClassName: '',
  headerColumnClassName: '',
  rows: [],
  rowClassName: '',
  stickHeader: false,
};

export default Body;
