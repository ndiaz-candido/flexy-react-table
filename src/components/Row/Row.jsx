import React, { Component } from 'react';
import PropTypes from 'prop-types';
import idGenerator from 'react-id-generator';
import { ScrollSection, RightSection, LeftSection } from './sections';
import ScrollContext from '../ScrollContext';

const defaultRowStyle = {
  display: 'flex',
  minHeight: '50px',
};
const defaultCellStyle = width => ({
  width,
  flex: `0 0 ${width}`,
});

class Row extends Component {
  static DEFAULT_COLUMN_WIDTH = 50;

  state = {
    stickRightSection: [],
    stickLeftSection: [],
    scrollSection: [],
  };

  constructor(props) {
    super(props);

    this.rowId = idGenerator();
  }

  static getDerivedStateFromProps({ columns, cellClassName, rowData }) {
    const stickLeftSection = [];
    const stickRightSection = [];
    const scrollSection = [];

    columns.forEach(column => {
      const columnDataKey = column.dataKey;
      const columnWidth = column.width || Row.DEFAULT_COLUMN_WIDTH;
      const cellContent = rowData[columnDataKey];
      const cell = (
        <div className={cellClassName} style={defaultCellStyle(columnWidth)}>
          {cellContent}
        </div>
      );

      switch (column.stickAlign) {
        case 'right':
          stickRightSection.push(cell);
          break;
        case 'left':
          stickLeftSection.push(cell);
          break;
        default:
          scrollSection.push(cell);
          break;
      }
    });

    return {
      stickLeftSection,
      stickRightSection,
      scrollSection,
    };
  }

  render() {
    const { stickLeftSection, stickRightSection, scrollSection } = this.state;
    const { className } = this.props;

    return (
      <div className={className} style={defaultRowStyle}>
        <LeftSection cells={stickLeftSection} />
        <ScrollContext.Consumer>
          {({ handleOnScroll, registerScrollSection }) => (
            <ScrollSection
              cells={scrollSection}
              onScroll={handleOnScroll}
              registerScrollSection={registerScrollSection}
              rowId={this.rowId}
            />
          )}
        </ScrollContext.Consumer>
        <RightSection cells={stickRightSection} />
      </div>
    );
  }
}

Row.propTypes = {
  className: PropTypes.any,
  cellClassName: PropTypes.any,
  columns: PropTypes.arrayOf(
    PropTypes.shape({
      dataKey: PropTypes.string,
      stickAlign: PropTypes.string,
      width: PropTypes.string,
    })
  ),
  rowData: PropTypes.object,
};

Row.defaultProps = {
  className: '',
  cellClassName: '',
  columns: [],
  rowData: {},
};

export default Row;
