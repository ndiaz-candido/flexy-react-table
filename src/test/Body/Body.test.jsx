import React from 'react';
import { shallow } from 'enzyme';
import Body from '../../components/Body';
import Row from '../../components/Row';

describe('Body', () => {
  const columns = [
    {
      dataKey: 'a',
      label: 'a',
    },
    {
      dataKey: 'b',
      label: 'b',
    },
  ];

  test('should render its default content', () => {
    const BodyComponent = shallow(<Body />);

    expect(BodyComponent).toMatchSnapshot();
  });

  test('should render a BaseRow for each declared Row in the table', () => {
    const rows = [<Row key="a" />, <Row key="b" />, <Row key="c" />];
    const BodyComponent = shallow(<Body rows={rows} />);

    expect(BodyComponent).toMatchSnapshot();
  });

  test('should stick the header when stickHeader is provided', () => {
    const BodyComponent = shallow(<Body stickHeader />);

    expect(BodyComponent).toMatchSnapshot();
  });

  test('should update the scroll left position when handleScrollEvent is called', () => {
    const BodyComponent = shallow(<Body />);
    const HeaderRow = BodyComponent.find('BaseRow').first();

    HeaderRow.prop('onScroll')({ scrollLeft: 10 });

    expect(BodyComponent.state()).toEqual({
      scrollLeft: 10,
    });
  });

  test('should pass the column labels to the header row if not renderer is provided', () => {
    const BodyComponent = shallow(<Body columns={columns} />);

    expect(BodyComponent).toMatchSnapshot();
  });

  test('should call the provided column header renderer', () => {
    const columnHeaderRenderer = ({ label }) => <div>{label}</div>;
    const BodyComponent = shallow(
      <Body columns={columns} columnHeaderRenderer={columnHeaderRenderer} />
    );

    expect(BodyComponent).toMatchSnapshot();
  });
});
