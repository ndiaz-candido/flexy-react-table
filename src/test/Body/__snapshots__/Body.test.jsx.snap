// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Body should call the provided column header renderer 1`] = `
<React.Fragment>
  <BaseRow
    className=""
    columnClassName=""
    columns={
      Array [
        Object {
          "dataKey": "a",
          "label": "a",
        },
        Object {
          "dataKey": "b",
          "label": "b",
        },
      ]
    }
    isHeader={true}
    isSticky={false}
    onScroll={[Function]}
    rowData={
      Object {
        "a": <div>
          a
        </div>,
        "b": <div>
          b
        </div>,
      }
    }
    rowId={0}
    scrollLeft={0}
  />
</React.Fragment>
`;

exports[`Body should pass the column labels to the header row if not renderer is provided 1`] = `
<React.Fragment>
  <BaseRow
    className=""
    columnClassName=""
    columns={
      Array [
        Object {
          "dataKey": "a",
          "label": "a",
        },
        Object {
          "dataKey": "b",
          "label": "b",
        },
      ]
    }
    isHeader={true}
    isSticky={false}
    onScroll={[Function]}
    rowData={
      Object {
        "a": "a",
        "b": "b",
      }
    }
    rowId={0}
    scrollLeft={0}
  />
</React.Fragment>
`;

exports[`Body should render a BaseRow for each declared Row in the table 1`] = `
<React.Fragment>
  <BaseRow
    className=""
    columnClassName=""
    columns={Array []}
    isHeader={true}
    isSticky={false}
    onScroll={[Function]}
    rowData={Object {}}
    rowId={0}
    scrollLeft={0}
  />
  <BaseRow
    className=""
    columnClassName=""
    columns={Array []}
    isHeader={false}
    isSticky={false}
    key="syncScrollRow-0"
    onScroll={[Function]}
    rowData={Object {}}
    rowId={1}
    scrollLeft={0}
  />
  <BaseRow
    className=""
    columnClassName=""
    columns={Array []}
    isHeader={false}
    isSticky={false}
    key="syncScrollRow-1"
    onScroll={[Function]}
    rowData={Object {}}
    rowId={2}
    scrollLeft={0}
  />
  <BaseRow
    className=""
    columnClassName=""
    columns={Array []}
    isHeader={false}
    isSticky={false}
    key="syncScrollRow-2"
    onScroll={[Function]}
    rowData={Object {}}
    rowId={3}
    scrollLeft={0}
  />
</React.Fragment>
`;

exports[`Body should render its default content 1`] = `
<React.Fragment>
  <BaseRow
    className=""
    columnClassName=""
    columns={Array []}
    isHeader={true}
    isSticky={false}
    onScroll={[Function]}
    rowData={Object {}}
    rowId={0}
    scrollLeft={0}
  />
</React.Fragment>
`;

exports[`Body should stick the header when stickHeader is provided 1`] = `
<React.Fragment>
  <BaseRow
    className=""
    columnClassName=""
    columns={Array []}
    isHeader={true}
    isSticky={true}
    onScroll={[Function]}
    rowData={Object {}}
    rowId={0}
    scrollLeft={0}
  />
</React.Fragment>
`;
